rfcs/20210519-single-pass-bnorm/README.md

```diff
@@ -0,0 +1,126 @@
+```
+Var(X) = E(X^2) - (E(X))^2
+```
+where X = input tensor, E = average
+
+This formula allows hardware to calculate both mean and variance in single pass
+over input tensor.
+For reference, the canonical formula is:
+```
+Var(X) = E((X - Mean(X))^2)
+```
+The benefit of new formula is 25-33% performance improvement. Batchnorm kernel
+is memory-bound, with new formula it will need to read input tensor 2 times 
+(1: mean&var, 2: normalize) instead of 3 times (1: mean, 2: var, 3: normalize)
+
+This formula will be used whenever batchnorm kernel calculates mean and 
+variance, which means forward propagation in training. However, with 
+non-standard combination of flags, oneDNN's batchnorm implementaton may be
+configured to calculate those two values in inference as well.
+
+## Design
+Algorithm for calculating variance will be implemented as:
+```
+Sum = 0
+SumOfSquares = 0
+for each x in X:
+	Sum += x
+	SumOfSquares += (x*x)
+Variance = SumOfSquares/N - (Sum/N)^2
+if Variance < 0 then Variance = 0
+```
+where N = number of items in given channel in tensor.
+
+In the formula there's a subtraction of two large values which is vulnerable
+to catastrophic cancellation. Rounding errors in the two big numbers may be
+large in relation to result of the formula. In order to minimize rounding
+errors:
+- Sum and SumOfSquares will be stored in fp32 precision, regardless of input
+	data type
+- Kahan summation algorithm will be used
+
+The last line, "if var < 0 then var = 0" is there to make sure rounding errors
+won't cause variance to go negative. Further in batchnorm calculations square
+root is taken from variance. Negative value would result in NaN.
+
+New formula should be considered experimental. By default the old one will be
+used. There needs to be a way for user to choose formula. It is not decided yet
+how exactly user will toggle it.
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_BNORM_ONE_PASS`. If not set or equal to zero, 
+old formula will be used.
+
+Pros:
+- API/ABI compatibility
+- Simplest to implement and use
+- Clearly defined default state
+- If needed, may be kept confidential and not disclosed to external customes 
+until tested internally
+
+Cons:
+- No fine-tuning per instance of bnorm primitive
+- Harder to reproduce bugs related to bnorm: value of this flag will need to be
+ reported in bug reports to allow reproduction.
+ - Inconvenient long-term. State of oneDNN will have a dependency which is not
+ obvious to users.
+
+### Enable through additional flag in bnorm primitive descriptor
+Bnorm descriptor has flags field `dnnl_normalization_flags_t`. Add new flag 
+`dnnl_one_pass = 0x20u`. When set to non-zero, new formula will be used.
+
+Pros:
+- Fine-tuning per instance of bnorm primitive
+- Suitable to become permanent solution, should both formulas be kept in oneDNN
+
+Cons:
+- API change needed.
+- In case a customer was passing garbage into bit5 of flags, the default
```

Comment:
This is a user error, not a cons of a proposed solution. User also can set `DNNL_BNORM_ONE_PASS` to a garbage, so there is no differences.